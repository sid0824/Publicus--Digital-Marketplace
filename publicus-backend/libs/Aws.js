import AWS from 'aws-sdk';
import {AccessKeyId, SecretAccessKey} from "../config/ConfigValues";

export const s3 = new AWS.S3({
  region: 'ap-southeast-2',
  accessKeyId: AccessKeyId,
  secretAccessKey: SecretAccessKey
});

export const ses = new AWS.SES({
  region: 'us-west-2'
});