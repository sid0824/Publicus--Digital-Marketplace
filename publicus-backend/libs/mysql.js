import mysql from 'mysql';
import {MySqlHost, MySqlPassword, MySqlUser} from "../config/ConfigValues";

//mysql connection to database
const connection = mysql.createConnection({
    host: MySqlHost,
    user: MySqlUser,
    password: MySqlPassword,
    database: "survey",
    multipleStatements: true
});

export const runQuery = async (query) => {
    return new Promise((resolve, reject) => {
        connection.query(query, (error, results) => {
            if(error) {
                reject(error);
            } else {
                resolve(results);
            }
        });
    });
};