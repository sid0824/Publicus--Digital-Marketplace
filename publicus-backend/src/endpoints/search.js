import {runQuery} from "../../libs/mysql";
import {failure, json_success} from "../util/ResponseUtil";

//Generates SQL Query (Returns Survey, Number of comments on survey and average Rating
//from surveys where either the name or one of the tags match the search term,
//with pagination of 10 results per page)
const selectSearchQuery = (page, search) => {
  return `SELECT s.*, u.avatar, u.firstname, u.lastname, COUNT(c.commentId) AS 'numComments', AVG(r.rating) AS 'avgRating' 
           FROM Surveys s 
           LEFT JOIN Comments c ON s.surveyId = c.surveyId 
           LEFT JOIN Users u ON u.userId = s.userId 
           LEFT JOIN Ratings r ON s.surveyId = r.surveyId 
           WHERE s.publishedStatus = 3 AND ( 
              s.name = '${search}' OR s.userId = '${search}' OR s.surveyId IN (
                SELECT surveyId
                FROM Tags
                WHERE tag = '${search}'
              )
           ) GROUP BY s.surveyId 
           LIMIT 10 OFFSET ${(page - 1) * 10}`;
};

// Generates SQL Query (Returns Survey, Number of comments on survey and average Rating
// from all surveys, with pagination of 10 results per page)
const selectAllSurveysQuery = (page) => {
  return `SELECT s.*, u.avatar, u.firstname, u.lastname, COUNT(c.commentId) AS 'numComments', AVG(r.rating) AS 'avgRating' 
         FROM Surveys s 
         LEFT JOIN Comments c ON s.surveyId = c.surveyId 
         LEFT JOIN Users u ON u.userId = s.userId 
         LEFT JOIN Ratings r ON s.surveyId = r.surveyId 
         WHERE s.publishedStatus = 3
         GROUP BY s.surveyId 
         LIMIT 10 OFFSET ${(page - 1) * 10}`;
};

//Generates SQL Query (Returns a list of tags that a surveyId has)
const selectTagQuery = (surveyId) => {
  return `SELECT tag 
          FROM Tags
          WHERE surveyId = '${surveyId}'`;
};

export async function main(event, context) {
    try {
      //Parse Page number from queryString (default is 1)
      const page = (event.queryStringParameters && event.queryStringParameters.page && parseInt(event.queryStringParameters.page)) || 1;

      //Parse whether a search parameter was given
      const searchParameterGiven = (event.queryStringParameters && event.queryStringParameters.search);

      //Select which query to run based on whether a search parameter was given
      let query;
      if(searchParameterGiven) {
        query = selectSearchQuery(page, searchParameterGiven);
      } else {
        query = selectAllSurveysQuery(page);
      }

      const results = await runQuery(query);

      const withTags = await Promise.all(results.map(async(result) => {
        //Find the number of tags for each survey
        //Then append those tags to that survey result
        const tagSql = selectTagQuery(result.surveyId);
        const tags = await runQuery(tagSql);

        //Modify each survey result to also include a list of tags for that survey
        return {
          ...result,
          tags
        };
      }));
      return json_success(withTags);
    } catch (err) {
      console.error(err);
      return failure();
    }
}
