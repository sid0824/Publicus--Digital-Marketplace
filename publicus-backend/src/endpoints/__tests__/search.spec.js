import { main } from '../search';
import { runQuery } from "../../../libs/mysql";

jest.mock("../../../libs/mysql");

const mockEvent = {
  requestContext: {
    identity: {
      cognitoIdentityId: 'MOCK_COGNITO_ID'
    },
  },
  body: JSON.stringify({
    comment: 'MOCK_COMMENT'
  }),
  queryStringParameters: {
    page: "2",
    search: 'SEARCH'
  }
};

describe('search', () => {

  beforeEach(() => {
  });

  afterEach(() => {
    runQuery.mockReset();
  });

  test(`should respond with failure when an error occurs`, async () => {
    runQuery.mockImplementation(() => {
      throw new Error('ERROR')
    });
    const response = await main(mockEvent);
    expect(response).toMatchSnapshot();
  });

  test(`should return forms with search from the database`, async () => {
    runQuery
      .mockResolvedValueOnce([{ surveyId: 'SURVEY_ID' }])
      .mockResolvedValueOnce([{ tag: 'TAG' }]);
    const response = await main(mockEvent);
    expect(runQuery.mock.calls[0][0]).toMatchSnapshot();
    expect(runQuery.mock.calls[1][0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  });

  test(`should return forms without search from the database`, async () => {
    runQuery
      .mockResolvedValueOnce([{ surveyId: 'SURVEY_ID' }])
      .mockResolvedValueOnce([{ tag: 'TAG' }]);
    const response = await main({ ...mockEvent,  queryStringParameters: {
        page: "2",
      }}
    );
    expect(runQuery.mock.calls[0][0]).toMatchSnapshot();
    expect(runQuery.mock.calls[1][0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  })

});