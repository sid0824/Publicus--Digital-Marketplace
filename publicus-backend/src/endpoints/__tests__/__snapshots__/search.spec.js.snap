// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search should respond with failure when an error occurs 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 500,
}
`;

exports[`search should return forms with search from the database 1`] = `
"SELECT s.*, u.avatar, u.firstname, u.lastname, COUNT(c.commentId) AS 'numComments', AVG(r.rating) AS 'avgRating' 
           FROM Surveys s 
           LEFT JOIN Comments c ON s.surveyId = c.surveyId 
           LEFT JOIN Users u ON u.userId = s.userId 
           LEFT JOIN Ratings r ON s.surveyId = r.surveyId 
           WHERE s.publishedStatus = 3 AND ( 
              s.name = 'SEARCH' OR s.userId = 'SEARCH' OR s.surveyId IN (
                SELECT surveyId
                FROM Tags
                WHERE tag = 'SEARCH'
              )
           ) GROUP BY s.surveyId 
           LIMIT 10 OFFSET 10"
`;

exports[`search should return forms with search from the database 2`] = `
"SELECT tag 
          FROM Tags
          WHERE surveyId = 'SURVEY_ID'"
`;

exports[`search should return forms with search from the database 3`] = `
Object {
  "body": "[{\\"surveyId\\":\\"SURVEY_ID\\",\\"tags\\":[{\\"tag\\":\\"TAG\\"}]}]",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "applications/json",
  },
  "statusCode": 200,
}
`;

exports[`search should return forms without search from the database 1`] = `
"SELECT s.*, u.avatar, u.firstname, u.lastname, COUNT(c.commentId) AS 'numComments', AVG(r.rating) AS 'avgRating' 
         FROM Surveys s 
         LEFT JOIN Comments c ON s.surveyId = c.surveyId 
         LEFT JOIN Users u ON u.userId = s.userId 
         LEFT JOIN Ratings r ON s.surveyId = r.surveyId 
         WHERE s.publishedStatus = 3
         GROUP BY s.surveyId 
         LIMIT 10 OFFSET 10"
`;

exports[`search should return forms without search from the database 2`] = `
"SELECT tag 
          FROM Tags
          WHERE surveyId = 'SURVEY_ID'"
`;

exports[`search should return forms without search from the database 3`] = `
Object {
  "body": "[{\\"surveyId\\":\\"SURVEY_ID\\",\\"tags\\":[{\\"tag\\":\\"TAG\\"}]}]",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "applications/json",
  },
  "statusCode": 200,
}
`;
